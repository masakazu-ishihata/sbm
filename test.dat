2
10
10
2
0 0 0
0 2 1
0 4 1
0 6 1
0 8 1
1 1 1
1 4 0
1 9 0
2 3 0
2 4 0
2 9 1
3 2 0
3 9 1
4 1 0
4 2 0
4 3 0
5 0 1
5 5 0
5 9 1
6 2 1
6 7 0
6 8 0
7 4 0
7 6 1
7 6 1
8 0 1
8 1 1
8 2 1
8 8 0
8 9 0
9 0 1
9 3 0
9 5 1
9 7 1
9 9 1