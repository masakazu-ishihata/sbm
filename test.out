lp = -91.15883102861561
z[0] = {1, 0, 0, 0, 1, 1, 0, 0, 1, 0}
z[1] = {0, 0, 0, 1, 1, 1, 1, 1, 1, 0}
{0, 0}, {0.6625514403292181, 0.08641975308641975, 0.2510288065843621}
{0, 1}, {0.7190082644628099, 0.1955922865013774, 0.08539944903581266}
{1, 0}, {0.3742331288343559, 0.2515337423312884, 0.3742331288343559}
{1, 1}, {0.7448559670781892, 0.12757201646090532, 0.12757201646090532}
