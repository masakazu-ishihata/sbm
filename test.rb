#!/usr/bin/env ruby

################################################################################
# default
################################################################################
@d = 2
@n = 3
@m = 30
@r = 1

################################################################################
# Arguments
################################################################################
require "optparse"
OptionParser.new { |opts|
  # options
  opts.on("-h","--help","Show this message") {
    puts opts
    exit
  }
  opts.on("-d [int]"){ |f|
    @d = f.to_i
  }
  opts.on("-n [int]"){ |f|
    @n = f.to_i
  }
  opts.on("-m [int]"){ |f|
    @m = f.to_i
  }
  opts.on("-r [int]"){ |f|
    @r = f.to_i
  }
  # parse
  opts.parse!(ARGV)
}

################################################################################
# Class
################################################################################
class MyTestData
  def initialize(_d, _n, _m, _r)
    @d = _d
    @n = _n
    @m = _m
    @r = _r
  end

  def generate
    #### header ####
    puts "#{@d}"
    for d in 0..@d-1
      puts "#{@m}"
    end
    puts "#{@r}"

    #### randomize clusters ####
    @c = Array.new(@d){|d| Array.new(@m){|i| rand(@n)} }

    #### randomize relations ####
    @R = Hash.new
    key = Array.new(@d){|d| 0}
    begin
      @R[key.clone] = rand(@r+1)
    end while (key = succ_key(key, @n)) != nil

    #### show data matrix ####
    key = Array.new(@d){|d| 0}
    begin
      ckey = Array.new(@d){|d| @c[d][ key[d] ]}
      puts "#{key.join(" ")} #{@R[ckey]-1}" if @R[ckey] > 0
    end while (key = succ_key(key, @m)) != nil
  end

  def succ_key(key, n)
    for d in 0..@d-1
      if key[d] < n-1
        key[d] += 1
        return key
      else
        key[d] = 0
      end
    end

    return nil
  end
end

################################################################################
# main
################################################################################
d = MyTestData.new(@d, @n, @m, @r)
d.generate
