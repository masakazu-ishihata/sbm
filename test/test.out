lp = -437.5091841451689
z[0] = {1, 1, 1, 2, 2, 0, 2, 0, 2, 2, 0, 2, 1, 1, 0, 0, 2, 2, 2, 0, 1, 0, 2, 1, 2, 1, 2, 2, 2, 2}
z[1] = {1, 2, 1, 1, 1, 0, 0, 2, 0, 1, 2, 1, 1, 2, 2, 1, 2, 0, 2, 1, 2, 2, 2, 2, 0, 0, 0, 0, 1, 1}
{0, 0}, {0.9827586206896551, 0.017241379310344827}
{0, 1}, {0.012658227848101266, 0.9873417721518988}
{0, 2}, {0.9873417721518988, 0.012658227848101266}
{1, 0}, {0.015151515151515152, 0.9848484848484849}
{1, 1}, {0.9888888888888889, 0.011111111111111112}
{1, 2}, {0.011111111111111112, 0.9888888888888889}
{2, 0}, {0.9918032786885246, 0.00819672131147541}
{2, 1}, {0.005988023952095809, 0.9940119760479041}
{2, 2}, {0.005988023952095809, 0.9940119760479041}
